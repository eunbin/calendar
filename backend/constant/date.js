export const dateFormat = Object.freeze({
  DATE: 'YYYY-MM-DD',
  DATE_TIME: 'YYYY-MM-DD HH:mm',
  HOUR_AND_MIN: 'HH:mm'
})
