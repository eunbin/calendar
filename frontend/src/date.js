export const dateFormat = Object.freeze({
  DATE: 'YYYY-MM-DD',
  DATE_TIME: 'YYYY-MM-DD HH:mm',
  HOUR_AND_MIN: 'HH:mm',
  YEAR_AND_MONTH: 'YYYY [년] M [월]',
  TIME: 'LT',
  DATE_AND_DAY: 'DD (dd)'
})
