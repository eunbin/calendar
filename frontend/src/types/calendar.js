const MONTH = 'month'
const WEEK = 'week'

const viewTypes = Object.freeze({
  MONTH,
  WEEK
})

export {
  viewTypes
}
