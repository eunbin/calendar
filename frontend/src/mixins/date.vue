<script>
import { dateFormat } from '@/date'

export default {
  data () {
    return {
      dateFormat
    }
  }
}
</script>
