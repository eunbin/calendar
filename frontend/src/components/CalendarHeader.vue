<template>
  <header>
    <div class="nav">
      <button @click="prev">
        이전
      </button>
      <h1>{{ title }}</h1>
      <button @click="next">
        다음
      </button>
    </div>
    <div class="view-type">
      <button
        v-for="(obj, index) in viewTypeList"
        :key="index"
        :class="{ active: obj.value === type }"
        @click="setViewType(obj.value)"
      >
        {{ obj.label }}
      </button>
    </div>
  </header>
</template>

<script>
import { viewTypes } from '@/types/calendar'

export default {
  name: 'CalendarHeader',
  props: {
    viewType: {
      type: String,
      default: ''
    },
    currentDate: {
      type: Object,
      default: null
    }
  },
  data () {
    return {
      type: this.viewType
    }
  },
  computed: {
    title () {
      return this.currentDate.format(this.dateFormat.YEAR_AND_MONTH)
    },
    viewTypeList () {
      return [{
        value: viewTypes.MONTH,
        label: '월'
      }, {
        value: viewTypes.WEEK,
        label: '주'
      }]
    }
  },
  methods: {
    setViewType (newType) {
      this.type = newType
      this.$emit('change', this.type)
    },
    prev () {
      this.$emit('prev')
    },
    next () {
      this.$emit('next')
    }
  }
}
</script>

<style lang="scss" scoped>
$header-height: 50px;

@mixin btn--normal {
  background-color: var(--color-gray);
  width: 100px;
  height: 100%;
  border: none;
  cursor: pointer;
}

header {
  height: $header-height;
  display: flex;
  justify-content: center;
  color: var(--color-primary);

  .nav {
    display: flex;
    align-items: center;
    & button {
      @include btn--normal;
    }
    & h1 {
      font-size: 2em;
      margin: 0 0.3em 0 0.3em;
    }
  }

  .view-type {
    margin-left: 50px;
    button {
      @include btn--normal;
      &:not(:last-child) {
        border-right: 1px solid #fff;
      }
      &:hover, &.active {
        background-color: var(--color-secondary);
      }
    }
  }
}
</style>
