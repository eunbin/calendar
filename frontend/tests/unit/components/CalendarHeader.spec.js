import { shallowMount, mount } from '@vue/test-utils'
import CalendarHeader from '@/components/CalendarHeader.vue'
import dateMixin from '@/mixins/date'
import { dateFormat } from '@/types/date'
import { viewTypes, DEFAULT_VIEW_TYPE } from '@/types/calendar'
import moment from 'moment'

const currentDate = moment().hours(0).minutes(0).seconds(0).milliseconds(0).clone()

const createOption = () => {
  return {
    mixins: [dateMixin],
    propsData: {
      currentDate,
      viewType: DEFAULT_VIEW_TYPE
    }
  }
}

describe('CalendarHeader.vue', () => {
  test('항상 동일한 컨텐츠가 렌더링되어야 한다.', () => {
    const wrapper = mount(CalendarHeader, createOption())
    expect(wrapper.vm.$el).toMatchSnapshot()
  })

  it('오늘 날짜를 기준으로 YYYY년 MM월 타이틀이 보여진다.', () => {
    const option = createOption()

    const wrapper = shallowMount(CalendarHeader, option)

    expect(wrapper.vm.title).toMatch(currentDate.format(dateFormat.YEAR_AND_MONTH))
  })

  it('기본 보기모드는 DEFAULT_VIEW_TYPE 과 같아야 한다.', () => {
    const option = createOption()

    const wrapper = shallowMount(CalendarHeader, option)

    expect(wrapper.vm.$data.type).toBe(DEFAULT_VIEW_TYPE)
  })

  it('제공하는 모든 뷰 타입에 대한 버튼이 생성된다.', () => {
    const option = createOption()

    const wrapper = mount(CalendarHeader, option)
    const buttonWrapper = wrapper.find('.view-type').findAll('button')

    expect(buttonWrapper.length).toBe(Object.keys(viewTypes).length)
  })

  it('<, > 이동 버튼의 클릭 이벤트가 정상 동작한다.', () => {
    const option = createOption()

    const wrapper = mount(CalendarHeader, option)

    const buttonWrapper = wrapper.find('.nav').findAll('button')
    const firstButton = buttonWrapper.at(0)
    expect(firstButton.is('button')).toBe(true)
    const secondButton = buttonWrapper.at(1)
    expect(secondButton.is('button')).toBe(true)

    firstButton.trigger('click')
    secondButton.trigger('click')

    expect(wrapper.emitted().prev).toBeTruthy()
    expect(wrapper.emitted().next).toBeTruthy()
  })

  it('기본 view type인 month 에서 view 로 변경되도록 값을 전달한다.', () => {
    const option = createOption()

    const wrapper = mount(CalendarHeader, option)

    const buttonWrapper = wrapper.find('.view-type').findAll('button')
    const secondButton = buttonWrapper.at(1)
    expect(secondButton.is('button')).toBe(true)

    secondButton.trigger('click')

    expect(wrapper.emitted().change[0][0]).toEqual('week')
  })
})
