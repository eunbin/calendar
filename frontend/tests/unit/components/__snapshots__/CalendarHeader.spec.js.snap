// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CalendarHeader should render content correctly 1`] = `
<header>
  <div
    class="nav"
  >
    <button>
      
      이전
    
    </button>
     
    <h1>
      2020 년 2 월
    </h1>
     
    <button>
      
      다음
    
    </button>
  </div>
   
  <div
    class="view-type"
  >
    <button
      class=""
    >
      <span
        class="title"
      >
        월
      </span>
    </button>
    <button
      class=""
    >
      <span
        class="title"
      >
        주
      </span>
    </button>
  </div>
</header>
`;

exports[`CalendarHeader 동일한 컨텐츠가 렌더링되어야 한다. 1`] = `
<header>
  <div
    class="nav"
  >
    <button>
      
      이전
    
    </button>
     
    <h1>
      2020 년 2 월
    </h1>
     
    <button>
      
      다음
    
    </button>
  </div>
   
  <div
    class="view-type"
  >
    <button
      class=""
    >
      
      월
    
    </button>
    <button
      class=""
    >
      
      주
    
    </button>
  </div>
</header>
`;

exports[`CalendarHeader 항상 동일한 컨텐츠가 렌더링되어야 한다. 1`] = `
<header>
  <div
    class="nav"
  >
    <button>
      
      이전
    
    </button>
     
    <h1>
      2020 년 2 월
    </h1>
     
    <button>
      
      다음
    
    </button>
  </div>
   
  <div
    class="view-type"
  >
    <button
      class=""
    >
      
      월
    
    </button>
    <button
      class=""
    >
      
      주
    
    </button>
  </div>
</header>
`;
